### PR TITLE
Address precedence issues in `where` patterns

### LINKED ISSUE #48: Precedence issue with type pattern and `where`
The following snippet demonstrates that a `when` condition is ignored (by Julia) in some circumstances.  In other circumstances it results in surprising errors.

Condition being ignored:
```julia
julia> cond = true
true

julia> @match2 10 begin
       x::Int where cond => 1
       end
1

julia> cond = false
false

julia> @match2 10 begin
       x::Int where cond => 1
       end
1

julia> @match2 10 begin
       (x::Int) where cond => 1
       end
ERROR: MatchFailure(10)
Stacktrace:
 [1] top-level scope
   @ REPL[6]:2

julia> Int where cond
Int64

julia> Int where zztop
Int64

julia>
```

Condition resulting in surprising errors:
```julia
julia> @match2 10 begin
       x::Int where cond && cond => 1
       end
1

julia> cond
false

julia> @match2 10 begin
       x::Int where (cond && cond) => 1
       end
ERROR: LoadError: REPL[14]:2: Could not bind `Int where cond && cond` as a type (due to `ErrorException("syntax: invalid variable expression in \"where\" around REPL[14]:2")`).
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] bind_pattern!(location::LineNumberNode, source::Expr, input::Symbol, state::Rematch2.BinderState, assigned::Base.ImmutableDict{Symbol, Symbol})
   @ Rematch2 ~/Repos/Rematch2.jl/src/BindPattern.jl:109
```

To address this, I propose to
1. Treat `x::T where a` in a pattern as `(x::T) where a`.
2. Give an error if the type `T` is syntactically anything other than a (optionally qualified) name, possibly with type arguments.

The disadvantage of doing this is that you would no longer be able to specify a type such as `Vector{T} where T <: Vector`, i.e., vectors of vectors.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #48

### CODE REVIEW

File: test/rematch2.jl [hunk 1]
```diff
@@ -1,13 +1,71 @@
...
+
+    x = false
+    @test (@match2 3 begin
+        ::Int where x => 1
+        _ => 2
+    end) == 2
+end
+
+@testset "Check that `where` clauses are reparsed properly 2" begin
+    x = true
```
Comment (mcmcgrath13) [inline test/rematch2.jl:26]: could you add a test along the lines of
```
x = 2
@test (@match2 3 begin
   a::Int where x == a => a
   _ => 2
end)
```
and the `!=` case as well? (or any falsey case)

**Review by mcmcgrath13 — APPROVED** [review]
Nice find! Definitely an odd corner of the syntax of Julia

File: test/rematch2.jl (hunk 1 above)
Comment (gafter) [reply test/rematch2.jl:26]: I have a sinking feeling it won't work right.  Maybe we should move back to the `pattern if condition end` syntax from `Match.jl`.  That syntax is unambiguous.

Comment (gafter) [reply test/rematch2.jl:26]: Huh.  It works.

**Pull request merged**